presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.OptimizerStatsRecorder;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class StatsRecordingPlanOptimizer
+        implements PlanOptimizer
+{
+    private final OptimizerStatsRecorder stats;
+
+    protected StatsRecordingPlanOptimizer(OptimizerStatsRecorder stats)
+    {
+        this.stats = requireNonNull(stats, "stats is null");
+        stats.register(this);
+    }
+
+    public final PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        PlanNode result;
+        long duration;
+        try {
+            long start = System.nanoTime();
+            result = doOptimize(plan, session, types, symbolAllocator, idAllocator);
+            duration = System.nanoTime() - start;
+        }
+        catch (RuntimeException e) {
+            stats.recordFailure(this);
+            throw e;
+        }
+        stats.record(this, duration);
```

Comment:
Does recording duration of `0` modifies stats in case of failure (I recall 0 might be meaningful data point because there is count metrics in `TimeDistribution`).

Reply:
in case of failure, we won't be here. The fail exc is rethrown.